rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,242 @@
+  - We have to draw a line as to what constitutes the public interface of Nixpkgs. We have decided that making attribute position information part of that is not productive. For context, this information is already accepted to be unreliable at the language level, noting the `unsafe` part of the name.
+  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Alternate `pkgs/unit` structure
+
+- Use a flat directory, e.g. `pkgs.hello` would be in `pkgs/unit/hello`.
+  - Good because it's simpler, both for the user and for the code
+  - Good because it speeds up Nix evaluation since there's only a single directory to call `builtins.readDir` on instead of many
+    - With an optimized `readDir` this isn't much of a problem
+  - Bad because the GitHub web interface only renders the first 1'000 entries (and we have about 10'000 that benefit from this transition, even given the restrictions)
+  - Bad because it makes `git` slower ([TODO: By how much?](https://github.com/nixpkgs-architecture/simple-package-paths/issues/18))
+  - Bad because directory listing slows down with many files
+- Use `substring 0 3 name` or `substring 0 4 name`. This was not done because it still leads to directories in `pkgs/unit` containing more than 1'000 entries, leading to the same problems.
+- Use multi-level structure, like a 2-level 2-prefix structure where `hello` is in `pkgs/unit/he/ll/hello`,
+  if packages are less than 4 characters long, we will it out with `-`, e.g. `z` is in `pkgs/unit/z-/--/z`.
+  This is not great because it's more complicated and it would improve git performance only marginally.
+- Use a dynamic structure where directories are rebalanced when they have too many entries.
+  E.g. `pkgs.foobar` could be in `pkgs/unit/f/foobar` initially.
+  But when there's more than 1'000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/unit/fo/foobar`.
+  This is not great because it's very complex to determine which directory to put a package in, making it bad for contributors.
+
+## Alternate `pkg-fun.nix` filename
+
+- `default.nix`: Bad because:
+  - Doesn't have its main benefits in this case:
+    - Removing the need to specify the file name in expressions, but this does not apply because this file will be imported automatically by the code that replaces definitions from `all-packages.nix`.
+    - Removing the need to specify the file name on the command line, but this does not apply because a package function must be imported into an expression before it can be used, making `nix build -f pkgs/unit/hell/hello` equally broken regardless of file name.
+  - Not using `default.nix` frees up `default.nix` for a short expression that is actually buildable, e.g. `(import ../..).hello`, although at that point it might better be auto-generated or implicit in the CLI
```

Comment:
good because it is consistent with everything so far, we don't need to explain to ~~new~~ contributors why those files are named different, muscle memory.

Reply:
I added the fact that it's an existing convention as an argument in https://github.com/NixOS/rfcs/commit/269ac4f669302c9b09075a40632b664158908a0b#diff-d795c40df63067333bb77ad1be78f1ea7e46350e124528338d970662a37f6b3bR152